### PR TITLE
Feat/loss soft hkr

### PR DESCRIPTION
**Pull request by franckma31**
Introduce the multiclass loss : MulticlassSoftHKR  defined in the paper "On the explainable properties of 1-Lipschitz Neural Networks: An Optimal Transport Perspective" Serrurier et al (Neurips'23) that combine an optimal transport loss with a softmax temperature coefficient (useful when the numer of classes is high like in Imagenet dataset)

### CODE REVIEW

**Review by cofri — CHANGES_REQUESTED** [review]

File: tests/test_losses.py [hunk 1]
```diff
@@ -174,6 +175,56 @@ def test_hkr_multiclass_loss(self):
...
+        multiclass_softhkr3 = MulticlassSoftHKR(5.0, 0.2, one_hot_ytrue=False)
+        loss_val_3 = multiclass_softhkr3(2 * y_true - 1, y_pred).numpy()
+        self.assertEqual(
+            loss_val_3,
+            loss_val,
+            "test failed when y_true has dtype +/-1",
+        )
+
+        check_serialization(1, multiclass_softhkr)
+
```
[the remark this answers is not in the document]
Comment (franckma31) [reply tests/test_losses.py:227]: test_minibatches_multiclass_losses is not possible dur to the moving_mean computed on the batch. Each sub minibatch would have a different mean and leading a a different temperature for the softmax. In real condition all moving_mean may converge to the same value but unit test will fail

test_noreduction will be ok

File: deel/lip/losses.py [hunk 2]
```diff
@@ -397,6 +397,170 @@ def get_config(self):
...
+        """
+        The multiclass version of HKR with softmax. This is done by computing
+        the HKR term over each class and averaging the results.
+
+        Note that `y_true` should be one-hot encoded or pre-processed with the
+        `deel.lip.utils.process_labels_for_multi_gpu()` function.
+
+        Using a multi-GPU/TPU strategy requires to set `multi_gpu` to True and to
+        pre-process the labels `y_true` with the
+        `deel.lip.utils.process_labels_for_multi_gpu()` function.
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/lip/losses.py:420]: comment on multi-gpu removed

File: deel/lip/losses.py [hunk 3]
```diff
@@ -397,6 +397,142 @@ def get_config(self):
...
+        self.alpha_mean = alpha_mean
+
+        self.current_mean = tf.Variable(
+            (self.min_margin_v,),
+            dtype=tf.float32,
+            constraint=lambda x: tf.clip_by_value(x, 0.005, 1000),
+            name="current_mean",
+        )
+
+        self.temperature = temperature * self.min_margin_v
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/lip/losses.py:439]: self.temperature remains constant during the training. The only variable is self.current_mean but the question of serialization of the value is good

File: deel/lip/losses.py [hunk 4]
```diff
@@ -397,6 +397,170 @@ def get_config(self):
...
+        self.alpha = tf.Variable(alpha, dtype=tf.float32)
+        self.min_margin_v = min_margin
+        self.moving_mean = None
+        self.alpha_mean = alpha_mean
+
+        self.current_mean = tf.Variable(
+            (self.min_margin_v,),
+            dtype=tf.float32,
+            constraint=lambda x: tf.clip_by_value(x, 0.005, 1000),
+            name="moving_mean",
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/lip/losses.py]: changed to current_mean in 413640dbfb7ee699a03bce3323b6c882c1cb4ab8

File: deel/lip/losses.py [hunk 5]
```diff
@@ -397,6 +397,170 @@ def get_config(self):
...
+        F_soft_KR = tf.where(y_true > 0, tf.cast(1.0, F_soft_KR.dtype), F_soft_KR)
+        # if self.stop_gradient:
+        #    F_soft_KR = tf.stop_gradient(F_soft_KR)
+        return F_soft_KR
+
+    def kr_preproc(self, y_true, y_pred):
+        """From _kr_multi_gpu(y_true, y_pred)"""
+        # assert tf.reduce_min(y_true)<0
+        if self.one_hot_ytrue:
+            y_true = tf.where(y_true > 0, 1, -1)  # switch to +/-1
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/lip/losses.py]: modification done in 691e74e0c22628a347ca218cbb79341bf2905273

File: deel/lip/losses.py [hunk 6]
```diff
@@ -397,6 +397,142 @@ def get_config(self):
...
+
+        Note that `y_true` could be either one-hot encoded, +/-1 values.
+
+
+        Args:
+            alpha (float): regularization factor
+            min_margin (float): margin to enforce.
+            temperature (float): factor for softmax  temperature
+                (higher value increases the weight of the highest non y_true logits)
+            alpha_mean (float): geometric mean factor
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/lip/losses.py]: done in 5d1d177b8f8ac2b93cc7742aa56a5c883d7aa6c4

File: deel/lip/losses.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/lip/losses.py:439]: resolved in 1b6205dda54d7fe9571b7617da9b9d4b04c65f87

**Review by cofri — APPROVED** [review]
👌
